src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,408 @@
+using SyntaxMap = System.Collections.Immutable.ImmutableDictionary<Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis.SyntaxNode[]>;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
+{
+    internal class MembersPuller
+    {
+        internal static readonly MembersPuller Instance = new MembersPuller();
+
+        private MembersPuller()
+        {
+        }
+
+        /// <summary>
+        /// Return the CodeAction to pull selectedMember up to destinationType. If the pulling will cause error,
+        /// it will return null.
+        /// </summary>
+        public CodeAction TryComputeCodeAction(
+            Document document,
+            ISymbol selectedMember,
+            INamedTypeSymbol destination)
+        {
+            var result = PullMembersUpAnalysisBuilder.BuildAnalysisResult(destination, ImmutableArray.Create((member: selectedMember, makeAbstract: false)));
+            if (result.PullUpOperationCausesError ||
+                IsSelectedMemberDeclarationAlreadyInDestination(selectedMember, destination))
+            {
+                return default;
+            }
+
+            return new SolutionChangeAction(
+                string.Format(FeaturesResources.Add_to_0, result.Destination),
+                cancellationToken => PullMembersUpAsync(document, result, cancellationToken));
+        }
+
+        /// <summary>
+        /// Return the changed solution if all changes in result are applied.
+        /// </summary>
+        /// <param name="result">Contains the members to pull up and all the fix operations</param>>
+        internal async Task<Solution> PullMembersUpAsync(
+            Document document,
+            PullMembersUpAnalysisResult result,
+            CancellationToken cancellationToken)
+        {
+            if (result.Destination.TypeKind == TypeKind.Interface)
+            {
+                return await PullMembersIntoInterfaceAsync(document, result, cancellationToken).ConfigureAwait(false);
+            }
+            else if (result.Destination.TypeKind == TypeKind.Class)
+            {
+                return await PullMembersIntoClassAsync(document, result, cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                throw ExceptionUtilities.UnexpectedValue(result.Destination);
+            }
+        }
+
+        private IMethodSymbol FilterGetterOrSetter(IMethodSymbol getterOrSetter)
+        {
+            // Filter the non-public getter/setter since the propery is going to pull up to an interface
+            return getterOrSetter?.DeclaredAccessibility == Accessibility.Public ? getterOrSetter : null;
+        }
+
+        private IMethodSymbol CreatePublicGetterAndSetter(IPropertySymbol containingProperty, IMethodSymbol getterOrSetter)
+        {
+            // Create a public getter/setter since user is trying to pull a non-public property to an interface.
```

Comment:
How is this relating to the prior function? We filter out non-public, but we also deal with it here?

Reply:
Remove the useless param